File: leetcode/src/search-in-rotated-sorted-array.cpp
class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int lo = 0;
        int hi = nums.size() - 1;
        int mid = (hi - lo) / 2 + lo;

        while (lo <= hi)
        {
            int mid = (hi - lo) / 2 + lo;
            if (nums[mid] == target)
            {
                return mid;
            }

            if (nums[lo] <= nums[mid] && nums[lo] <= target && target < nums[mid])
            {
                hi = mid - 1;
            }
            else if (nums[lo] <= nums[mid])
            {
                lo = mid + 1;
            }
            else if (nums[mid] < target && target <= nums[hi])
            {
                lo = mid + 1;
            }
            else
            {
                hi = mid - 1;
            }
        }

        return -1;
    }
};

File: leetcode/src/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:

        lo = 0
        hi = len(nums)-1


        while (lo <= hi):
            mid = (lo + hi)//2
            if (nums[mid] == target):
                return mid
            if (nums[lo] <= nums[mid]): #pivoted is not in lo and mid
                if (nums[lo] <= target and target <= nums[mid]):
                    hi = mid - 1
                else:
                    lo = mid + 1
                continue
            else: #pivot is not in mid and hi
                if (nums[mid] <= target and target <= nums[hi]):
                    lo = mid + 1
                else:
                    hi = mid - 1
                continue
        
        return -1


